Kernel/Graphics/VirtIOGPU/VirtIOGPU.cpp

```diff
@@ -0,0 +1,286 @@
+                for (auto i3 = 0; i3 < 2; ++i3) {
+                    auto i = width * (ny + (height / 2)) + nx + (width / 2);
+                    data[(4 * i) + 0] = 0;
+                    data[(4 * i) + 1] = 0;
+                    data[(4 * i) + 2] = 0xff;
+                    data[(4 * i) + 3] = 0xff;
+                    ny *= -1;
+                }
+                nx *= -1;
+            }
+        }
+    };
+
+    for (int y = 0; y < 200; ++y) {
+        for (int x = 0; x * x + y * y < (150 * 150); ++x) {
+            draw_red_pixel_mirrored(x, y);
+        }
+    }
+    dbgln_if(VIRTIO_DEBUG, "Finish drawing the pattern");
+}
+
+void VirtIOGPU::transfer_framebuffer_data_to_host(VirtIOGPURect dirty_rect)
+{
+    Locker locker(m_operation_lock);
+    auto& request = *reinterpret_cast<VirtIOGPUTransferToHost2D*>(m_scratch_space->vaddr().as_ptr());
+    auto& response = *reinterpret_cast<VirtIOGPUCtrlHeader*>((m_scratch_space->vaddr().offset(sizeof(request)).as_ptr()));
+
+    populate_virtio_gpu_request_header(request.header, VirtIOGPUCtrlType::VIRTIO_GPU_CMD_TRANSFER_TO_HOST_2D, VIRTIO_GPU_FLAG_FENCE);
+    request.offset = (dirty_rect.x + (dirty_rect.y * m_display_info.rect.width)) * sizeof(u32);
+    request.resource_id = m_framebuffer_id;
+    request.rect = dirty_rect;
+
+    synchronous_virtio_gpu_command(start_of_scratch_space(), sizeof(request), sizeof(response));
+
+    VERIFY(response.type == static_cast<u32>(VirtIOGPUCtrlType::VIRTIO_GPU_RESP_OK_NODATA));
+}
+
+void VirtIOGPU::flush_displayed_image(VirtIOGPURect dirty_rect)
+{
+    Locker locker(m_operation_lock);
+    auto& request = *reinterpret_cast<VirtIOGPUResourceFlush*>(m_scratch_space->vaddr().as_ptr());
+    auto& response = *reinterpret_cast<VirtIOGPUCtrlHeader*>((m_scratch_space->vaddr().offset(sizeof(request)).as_ptr()));
+
+    populate_virtio_gpu_request_header(request.header, VirtIOGPUCtrlType::VIRTIO_GPU_CMD_RESOURCE_FLUSH, VIRTIO_GPU_FLAG_FENCE);
+    request.resource_id = m_framebuffer_id;
+    request.rect = dirty_rect;
+
+    synchronous_virtio_gpu_command(start_of_scratch_space(), sizeof(request), sizeof(response));
+
+    VERIFY(response.type == static_cast<u32>(VirtIOGPUCtrlType::VIRTIO_GPU_RESP_OK_NODATA));
+}
+
+void VirtIOGPU::synchronous_virtio_gpu_command(PhysicalAddress buffer_start, size_t request_size, size_t response_size)
+{
+    VERIFY(!m_has_outstanding_request.load());
+    m_has_outstanding_request = true;
+    auto& queue = get_queue(CONTROLQ);
+    {
+        ScopedSpinLock lock(queue.lock());
+        VirtIOQueueChain chain { queue };
+        chain.add_buffer_to_chain(buffer_start, request_size, BufferType::DeviceReadable);
+        chain.add_buffer_to_chain(buffer_start.offset(request_size), response_size, BufferType::DeviceWritable);
+        supply_chain_and_notify(CONTROLQ, chain);
+        full_memory_barrier();
+    }
+
+    while (m_has_outstanding_request.load())
+        Scheduler::yield();
```

Comment:
can we just use a `WaitQueue` instead of reimplementive a primitive version of it ourselves?

Reply:
Fixed in next revision.